metadata/metadata.go

```diff
@@ -182,19 +182,46 @@ func FromIncomingContext(ctx context.Context) (MD, bool) {
 	if !ok {
 		return nil, false
 	}
-	out := MD{}
+	out := make(MD, len(md))
 	for k, v := range md {
 		// We need to manually convert all keys to lower case, because MD is a
 		// map, and there's no guarantee that the MD attached to the context is
 		// created using our helper functions.
 		key := strings.ToLower(k)
-		s := make([]string, len(v))
-		copy(s, v)
-		out[key] = s
+		out[key] = copyOf(v)
 	}
 	return out, true
 }
 
+// ValueFromIncomingContext returns the metadata value corresponding to the metadata
+// key from the incoming metadata if it exists. Key must be lower-case.
```

Comment:
Please add an experimental tag here so we can make sure this API works well in practice before considering it stable.

Example of our syntax used for them:

https://github.com/grpc/grpc-go/blob/1c29e075ab1c83a455cf942dd148c467a879800a/clientconn.go#L537-L540